File: day_6/puzzle.py
"""
AOC 2018, Day 6
"""

__author__ = "adibis"
__version__ = "0.1.0"
__license__ = "MIT"

def get_mhtn_dist(min_x, max_x, min_y, max_y, read_data, bounded_area):
    nodes = dict()
    bounded_region = 0

    # For all points on the canvas...
    for x in range(min_x, max_x):
        for y in range(min_y, max_y):
            min_dist = 999999
            eq_dist = False
            closest_node = ""
            total_dist = 0
            # For each node from the input list...
            for node in read_data:
                node = node.strip()
                node_x, node_y = node.split(',')
                node_x = int(node_x)
                node_y = int(node_y)
                # Calculate the distance of each node from each point on canvas
                this_dist = abs(node_x-x) + abs(node_y-y)
                # Keep storing total distance of this point from each node
                total_dist += this_dist

                if node not in nodes:
                    nodes[node] = 0
                # If the point is closer to a node than any previous node then it
                # needs to be marked properly. Keep checking this till you find the closest point.
                if(this_dist <= min_dist):
                    if(this_dist != min_dist):
                        min_dist = this_dist
                        eq_dist = False
                        closest_node = node
                    else:
                        eq_dist = True

            # Once you find the closest point then increase that point's influence by 1.
            # Our aim is to find point with max influence.
            if not eq_dist:
                nodes[closest_node] += 1

            # In the meantime, if a point is within a certain max_bound from each node then
            # add that point to the list of valid points.
            if total_dist < bounded_area:
                bounded_region += 1

    print (bounded_region)
    return nodes


def main():
    """Day six of AOC 2018"""
    with open ("input.txt", 'r') as f:
        read_data = f.readlines()

    # This is _NOT_ ideal. Python int being unbounded - use a value that works for this case.
    # It will not work for all cases, obviously.
    max_x = 0
    min_x = 10000
    max_y = 0
    min_y = 10000
    bounded_area = 10000

    # No point calculating distance for all random points.
    # Capture the bounds and use them alone.
    for code in read_data:
        x, y = code.split(',')
        x = int(x)
        y = int(y)
        if (x > max_x):
            max_x = x
        if (y > max_y):
            max_y = y
        if (x < min_x):
            min_x = x
        if (y < min_y):
            min_y = y

    dist_pass_one = get_mhtn_dist(min_x, max_x, min_y, max_y, read_data, bounded_area)
    dist_pass_two = get_mhtn_dist(min_x-1, max_x+1, min_y-1, max_y+1, read_data, bounded_area)

    # The unbounded points will change distance from a node if we increase the canvas size.
    # So - capture the distances on smallest canvas and smallest canvas + 1.
    # Find the distances that don't change between the two and that's the maximum.
    final_dict = dict()
    for key in dist_pass_one:
        if dist_pass_one[key] == dist_pass_two[key]:
            final_dict[key] = dist_pass_one[key]

    print (max(final_dict.values()))


if __name__ == "__main__":
    """ This is executed when run from the command line """
    main()

File: day_6/puzzle.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>
#include<string.h>
#include<limits.h>

#define REALLOC_SIZE 32

typedef struct {
    int x;
    int y;
    int influence;
} node_t;

void setup_max_coord(int *min_x, int *max_x, int *min_y, int *max_y, const node_t node) {

    if (*min_x > node.x)
        *min_x = node.x;
    if (*min_y > node.y)
        *min_y = node.y;
    if (*max_x < node.x)
        *max_x = node.x;
    if (*max_y < node.y)
        *max_y = node.y;
}

void gen_dist_values(int min_x, int min_y, int max_x, int max_y, int total_nodes, node_t **nodes) {
#ifdef TEST
    int bounded_area = 32;
#else
    int bounded_area = 10000;
#endif
    int bounded_region = 0;
    for (int x = min_x; x < max_x; x++) {
        for (int y = min_y; y < max_y; y++) {
            int min_dist = INT_MAX;
            int closest_node = -1;
            bool eq_dist = false;
            int this_dist = 0;
            int total_dist = 0;
            for (int i = 0; i < total_nodes; i++) {
                this_dist = abs((*nodes)[i].x - x) + abs((*nodes)[i].y - y);
#ifdef DEBUG
                printf("Dist for (x,y) = (%d,%d) for node[i] = node[%d]\n", x, y, i);
                printf("Dist = %d\n", this_dist);
#endif
                total_dist += this_dist;
                if (this_dist <= min_dist) {
                    if (this_dist != min_dist) {
                        min_dist = this_dist;
                        eq_dist = false;
                        closest_node = i;
                    } else {
                        eq_dist = true;
                    }
                }
            }
            if (eq_dist == false) {
                (*nodes)[closest_node].influence += 1;
            }
            if (total_dist < bounded_area) {
                bounded_region += 1;
            }
        }
    }

    printf("PART 2: Total bounded region = %d\n", bounded_region);
}

int main() {

    FILE *fp;
    size_t realloc_counter = REALLOC_SIZE;

    // Maximum X and Y coordinates seen in the input.
    int max_x = 0;
    int max_y = 0;
    int min_x = INT_MAX;
    int min_y = INT_MAX;

    node_t node;
    node_t *nodes_a = malloc(realloc_counter*sizeof(node_t));
    node_t *nodes_b = malloc(realloc_counter*sizeof(node_t));
    int total_nodes = 0;

    int max_influence = 0;

    // Read the file.
#ifdef TEST
    fp = fopen("./test.txt", "r");
#else
    fp = fopen("./input.txt", "r");
#endif
    if (fp == NULL) {
        perror("input.txt");
        exit(EXIT_FAILURE);
    }

    // Store all values in a buffer.
    while((fscanf(fp, "%d, %d\n", &node.x, &node.y)) != EOF) {

        setup_max_coord(&min_x, &max_x, &min_y, &max_y, node);
        node.influence = 0;
        nodes_a[total_nodes] = node;
        nodes_b[total_nodes] = node;
        total_nodes++;
#ifdef DEBUG
        printf("Adding node: %d, %d\n", node.x, node.y);
#endif
        if(total_nodes >= (int)realloc_counter) {
            realloc_counter += REALLOC_SIZE;
            nodes_a = realloc(nodes_a, realloc_counter*sizeof(node_t));
            nodes_b = realloc(nodes_b, realloc_counter*sizeof(node_t));
        }
    }
    fclose(fp);

#ifdef DEBUG
    printf("Coord: %d, %d\n", min_x, max_x);
    printf("Coord: %d, %d\n", min_y, max_y);
#endif

    gen_dist_values(min_x, min_y, max_x, max_y, total_nodes, &nodes_a);
    gen_dist_values(min_x-1, min_y-1, max_x+1, max_y+1, total_nodes, &nodes_b);

    for (int i = 0; i < total_nodes; i++) {
        if (nodes_a[i].influence == nodes_b[i].influence) {
            if(nodes_a[i].influence > max_influence) {
                max_influence = nodes_a[i].influence;
            }
        }
    }

    printf("PART 1: Max Influence = %d\n", max_influence);

    // Free all the allocated memory.
    free(nodes_a);
    free(nodes_b);

    return 0;
}
